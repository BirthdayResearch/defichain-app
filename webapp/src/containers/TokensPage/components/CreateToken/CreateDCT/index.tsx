import React from 'react';
import { Helmet } from 'react-helmet';
import { I18n } from 'react-redux-i18n';
import { NavLink } from 'react-router-dom';
import classnames from 'classnames';
import { MdArrowBack, MdCheck } from 'react-icons/md';
import {
  Button,
  Form,
  FormGroup,
  Label,
  Input,
  FormText,
  UncontrolledDropdown,
  DropdownToggle,
  DropdownMenu,
  DropdownItem,
  Row,
  Col,
} from 'reactstrap';

import Footer from './Footer';
import styles from './CreateDCT.module.scss';
import { TOKENS_PATH } from '../../../../../constants';
import { ITokenResponse } from '../../../../../utils/interfaces';
import Header from '../../../../HeaderComponent';
import {
  getPageTitle,
  getTransactionAddressLabel,
} from '../../../../../utils/utility';
import AddressDropdown from '../../../../../components/AddressDropdown';
import { CreateTokenFormState } from '..';

interface CreateDCTProps {
  handleChange: (e) => void;
  formState: CreateTokenFormState;
  IsCollateralAddressValid: boolean;
  isConfirmationModalOpen: string;
  setIsConfirmationModalOpen: (state: string) => void;
  createdTokenData: ITokenResponse;
  updatedTokenData: ITokenResponse;
  wait: number;
  setWait: (wait: number) => void;
  createConfirmation: () => void;
  updateConfirmation: () => void;
  handleDropDowns: (newData: any, amount: any) => void;
  cancelConfirmation: () => void;
  isErrorCreatingToken: string;
  isErrorUpdatingToken: string;
  isUpdate: boolean;
}

const getTransactionLabel = (formState: CreateTokenFormState) => {
  return getTransactionAddressLabel(
    formState.receiveLabel,
    formState.receiveAddress,
    I18n.t('containers.tokens.createToken.collateralAddress')
  );
};

const CreateDCT: React.FunctionComponent<CreateDCTProps> = (
  props: CreateDCTProps
) => {
  const {
    isUpdate,
    handleChange,
    formState,
    IsCollateralAddressValid,
    handleDropDowns,
    isConfirmationModalOpen,
    setIsConfirmationModalOpen,
    cancelConfirmation,
    createConfirmation,
    updateConfirmation,
    wait,
    createdTokenData,
    isErrorCreatingToken,
    updatedTokenData,
    isErrorUpdatingToken,
  } = props;

  return (
    <>
      <Helmet>
        <title>
          {getPageTitle(I18n.t('containers.tokens.tokensPage.title'))}
        </title>
      </Helmet>
      <Header>
        <Button
          to={TOKENS_PATH}
          tag={NavLink}
          color='link'
          className='header-bar-back'
        >
          <MdArrowBack />
          <span className='d-lg-inline'>
            {I18n.t('containers.tokens.createToken.back')}
          </span>
        </Button>
        <h1 className={classnames({ 'd-none': false })}>
          {isUpdate
            ? I18n.t('containers.tokens.createToken.updateTitle')
            : I18n.t('containers.tokens.createToken.createTitle')}
        </h1>
      </Header>
      <div className='content'>
        <section>
          <Form>
            <FormGroup className='form-label-group'>
              <Input
                type='text'
                placeholder={I18n.t('containers.tokens.createToken.name')}
                name='name'
                id='name'
                value={formState.name}
                onChange={handleChange}
                valid={formState.name.length > 0}
                invalid={formState.name.length > 128}
              />
              <Label for='message'>
                {I18n.t('containers.tokens.createToken.name')}
              </Label>
            </FormGroup>
            <FormGroup className='form-label-group'>
              <Input
                type='text'
                placeholder={I18n.t('containers.tokens.createToken.symbol')}
                name='symbol'
                id='symbol'
                value={formState.symbol}
                onChange={handleChange}
                required
                valid={formState.symbol.length > 0}
                invalid={formState.symbol.length > 8}
                disabled={isUpdate}
              />
              <Label for='message'>
                {I18n.t('containers.tokens.createToken.symbol')}
              </Label>
            </FormGroup>
            {/* NOTE: Do not remove, for future purpose */}
            {/* <FormGroup className='form-label-group'>
              <Input
                type='number'
                inputMode='numeric'
                placeholder={I18n.t('containers.tokens.createToken.decimal')}
                name='decimal'
                id='decimal'
                value={formState.decimal}
                onChange={handleChange}
              />
              <Label for='message'>
                {I18n.t('containers.tokens.createToken.decimal')}
              </Label>
              <FormText className='mt-2'>
                {I18n.t('containers.tokens.createToken.decimalText')}
              </FormText>
            </FormGroup>
            <FormGroup className='form-label-group'>
              <Input
                type='number'
                inputMode='numeric'
                placeholder={I18n.t('containers.tokens.createToken.limit')}
                name='limit'
                id='limit'
                value={formState.limit}
                onChange={handleChange}
              />
              <Label for='message'>
                {I18n.t('containers.tokens.createToken.limit')}
              </Label>
            </FormGroup>
            <FormGroup row>
              <Label sm={4} for='mintable'>
                <strong>
                  {I18n.t('containers.tokens.createToken.mintingSupport')}
                </strong>
              </Label>
              <Col sm={{ size: 8 }}>
                <FormGroup row>
                  <Col md='2'>
                    <FormGroup>
                      <CustomInput
                        type='radio'
                        name='mintable'
                        value={'false'}
                        id='mintable'
                        label={I18n.t('containers.tokens.createToken.no')}
                        checked={formState.mintable === 'false'}
                        onChange={handleChange}
                      />
                    </FormGroup>
                  </Col>
                  <Col>
                    <FormGroup>
                      <CustomInput
                        type='radio'
                        name='mintable'
                        value={'true'}
                        id='mintable2'
                        label={I18n.t('containers.tokens.createToken.yes')}
                        checked={formState.mintable === 'true'}
                        onChange={handleChange}
                      />
                    </FormGroup>
                  </Col>
                </FormGroup>
              </Col>
            </FormGroup>
            <FormText className='mb-5'>
              {I18n.t('containers.tokens.createToken.mintText')}
            </FormText>
            <Row>
              <Col md='4'>
                <FormGroup>
                  <Label>
                    <strong>
                      {I18n.t('containers.tokens.createToken.tradeable')}
                    </strong>
                  </Label>
                </FormGroup>
              </Col>
              <Col md='8' lg='6'>
                <Row className='mb-5'>
                  <Col md='2'>
                    <FormGroup>
                      <CustomInput
                        type='radio'
                        name='tradeable'
                        value={'false'}
                        id='tradeable1'
                        label={I18n.t('containers.tokens.createToken.no')}
                        checked={formState.tradeable === 'false'}
                        onChange={handleChange}
                      />
                    </FormGroup>
                  </Col>
                  <Col>
                    <FormGroup>
                      <CustomInput
                        type='radio'
                        name='tradeable'
                        value={'true'}
                        id='tradeable2'
                        label={I18n.t('containers.tokens.createToken.yes')}
                        checked={formState.tradeable === 'true'}
                        onChange={handleChange}
                      />
                    </FormGroup>
                  </Col>
                </Row>
              </Col>
            </Row> */}
            <AddressDropdown
              formState={formState}
              getTransactionLabel={getTransactionLabel}
              onSelectAddress={handleDropDowns}
              additionalClass={() =>
                !IsCollateralAddressValid ? styles.collateralDropdown : ''
              }
              isDisabled={isUpdate}
            />
            {!IsCollateralAddressValid && (
              <FormText className={`${styles.collateralFormText} mt-2`}>
                {I18n.t('containers.tokens.createToken.collateralAddressError')}
              </FormText>
            )}
          </Form>
        </section>
      </div>
      <Footer
        isUpdate={isUpdate}
        formState={formState}
        isConfirmationModalOpen={isConfirmationModalOpen}
        setIsConfirmationModalOpen={setIsConfirmationModalOpen}
        cancelConfirmation={cancelConfirmation}
        createConfirmation={createConfirmation}
        updateConfirmation={updateConfirmation}
        IsCollateralAddressValid={IsCollateralAddressValid}
        wait={wait}
        createdTokenData={createdTokenData}
        isErrorCreatingToken={isErrorCreatingToken}
        isErrorUpdatingToken={isErrorUpdatingToken}
        updatedTokenData={updatedTokenData}
      />
    </>
  );
};

export default CreateDCT;
