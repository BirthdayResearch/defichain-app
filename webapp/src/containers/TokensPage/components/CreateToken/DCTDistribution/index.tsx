// Not in use component. Don't delete it for future purposes

import React, { useState } from 'react';
import { Helmet } from 'react-helmet';
import { I18n } from 'react-redux-i18n';
import { NavLink } from 'react-router-dom';
import QrReader from 'react-qr-reader';
import { MdFileUpload, MdAdd, MdArrowBack } from 'react-icons/md';
import classnames from 'classnames';
import { Row, Col, ButtonGroup, Button, Modal, ModalBody } from 'reactstrap';

import CsvReader from '../../../../../components/CsvReader';
import Spinner from '../../../../../components/Svg/Spinner';
import AddressList from './AddressList';
import * as log from '../../../../../utils/electronLogger';
import { TOKENS_PATH, CREATE_DCT } from '../../../../../constants';
import Header from '../../../../HeaderComponent';
import { getPageTitle } from '../../../../../utils/utility';

interface DCTDistributionProps {
  handleActiveTab: (active: string) => void;
  csvData: any;
  setCsvData: (data) => void;
  handleSubmit: () => void;
  setIsVerifyingCollateralModalOpen: (state: boolean) => void;
  IsVerifyingCollateralModalOpen: boolean;
}

const DCTDistribution: React.FunctionComponent<DCTDistributionProps> = (
  props: DCTDistributionProps
) => {
  const [newAddressFlag, setNewAddressFlag] = React.useState<boolean>(false);
  const [flashed, setFlashed] = React.useState<string>('');
  const [uploadCsv, setuploadCsv] = useState<boolean>(false);
  const [openScanner, setOpenScanner] = useState<boolean>(false);

  const {
    handleActiveTab,
    setCsvData,
    csvData,
    handleSubmit,
    IsVerifyingCollateralModalOpen,
    setIsVerifyingCollateralModalOpen,
  } = props;

  const handleOnDrop = (data, file) => {
    if (file.type !== 'text/csv') {
      setuploadCsv(false);
    } else {
      const transformedData = data.map((address) => address.data[0]);
      setCsvData(transformedData);
      setuploadCsv(false);
    }
  };

  const handleOnEnterPress = (e) => {
    setCsvData([...csvData, e.target.value]);
  };

  const handleOnError = (err, file, inputElem, reason) => {
    setCsvData([]);
    setuploadCsv(false);
  };

  const handleOnRemoveFile = () => {
    setCsvData([]);
    setuploadCsv(false);
  };

  const handleDeleteAll = () => {
    setCsvData([]);
  };

  const handleDelete = (address: string) => {
    const filteredData = csvData.filter((addr) => addr !== address);
    setCsvData(filteredData);
  };

  const handleAddNewAddress = () => {
    setNewAddressFlag(true);
  };

  const handleOpenScanner = () => {
    setOpenScanner(true);
  };

  const handleToggleScanner = () => {
    setOpenScanner(false);
  };

  const handleScanError = (err) => {
    log.error(err);
  };

  const handleScan = (data) => {
    const updatedState = {
      flashed: 'flashed',
      toAddress: '',
      uriData: '',
    };
    if (data) {
      if (data.includes('DFI')) {
        updatedState.uriData = data;
      } else {
        updatedState.toAddress = data;
      }
      // shutterSnap.play();
      // this.setState(updatedState);
      // setTimeout(() => {
      //   this.isQRCodeValid();
      //   this.toggleScanner();
      //   this.setState({
      //     flashed: '',
      //   });
      // }, 600);
    }
  };

  return (
    <>
      <Helmet>
        <title>
          {getPageTitle(I18n.t('containers.tokens.tokensPage.title'))}
        </title>
      </Helmet>
      <Header>
        <Button
          onClick={() => handleActiveTab(CREATE_DCT)}
          color='link'
          className='header-bar-back'
        >
          <MdArrowBack />
          <span className='d-lg-inline'>
            {I18n.t('containers.tokens.dctDistribution.back')}
          </span>
        </Button>
        <h1 className={classnames({ 'd-none': false })}>
          {I18n.t('containers.tokens.dctDistribution.dctDistribution')}
        </h1>
        <ButtonGroup>
          <Button color='link' onClick={() => setuploadCsv(true)}>
            <MdFileUpload />
            <span className='d-lg-inline'>
              {I18n.t('containers.tokens.dctDistribution.uploadCsv')}
            </span>
          </Button>
          <Button color='link' onClick={() => handleAddNewAddress()}>
            <MdAdd />
            <span className='d-lg-inline'>
              {I18n.t('containers.tokens.dctDistribution.addAddress')}
            </span>
          </Button>
        </ButtonGroup>
      </Header>
      <div className='content'>
        <section className='h-100'>
          <AddressList
            csvData={csvData}
            handleDeleteAll={handleDeleteAll}
            handleDelete={handleDelete}
            newAddressFlag={newAddressFlag}
            openScanner={handleOpenScanner}
            handleOnEnterPress={handleOnEnterPress}
          />
          <Modal
            isOpen={uploadCsv}
            centered={true}
            toggle={() => setuploadCsv(false)}
          >
            <ModalBody className='p-5'>
              <CsvReader
                handleOnDrop={handleOnDrop}
                handleOnError={handleOnError}
                handleOnRemoveFile={handleOnRemoveFile}
              />
            </ModalBody>
          </Modal>
          <Modal
            isOpen={openScanner}
            toggle={handleToggleScanner}
            centered={true}
            className={`qr-scanner ${flashed}`}
          >
            <ModalBody>
              <QrReader
                delay={1000}
                onError={handleScanError}
                onScan={handleScan}
                showViewFinder={false}
                className='qr-scanner-preview w-100'
              />
            </ModalBody>
          </Modal>
        </section>
      </div>
      <footer className='footer-bar'>
        <div
          className={classnames({
            'd-none': IsVerifyingCollateralModalOpen,
          })}
        >
          <Row className='justify-content-between align-items-center'>
            <Col className='col-auto'>
              <div className='caption-secondary'>
                {I18n.t('containers.tokens.createToken.dfiRequired')}
              </div>
              <div>
                {'1,000'}
                &nbsp;
                {'DFI'}
              </div>
            </Col>
            <Col className='d-flex justify-content-end'>
              <Button
                to={TOKENS_PATH}
                tag={NavLink}
                color='link'
                className='mr-3'
              >
                {I18n.t('containers.tokens.createToken.cancel')}
              </Button>
              <Button
                color='primary'
                disabled={!csvData.length}
                onClick={() => {
                  setIsVerifyingCollateralModalOpen(true);
                  handleSubmit();
                }}
              >
                {I18n.t('containers.tokens.createToken.continue')}
              </Button>
            </Col>
          </Row>
        </div>
        <div
          className={classnames({
            'd-none': !IsVerifyingCollateralModalOpen,
          })}
        >
          <div className='footer-sheet'>
            <dl className='row'>
              <dd className='col-12'>
                <Spinner />
                <span className='mb-0'>
                  {I18n.t(
                    'containers.tokens.dctDistribution.verifyingCollateral'
                  )}
                </span>
              </dd>
            </dl>
          </div>
        </div>
      </footer>
    </>
  );
};

export default DCTDistribution;
