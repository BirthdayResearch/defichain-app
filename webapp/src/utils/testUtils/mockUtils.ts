import PersistentStore from '../persistentStore';
import * as isElectron from '../isElectron';
import * as Notifications from '../notifications';
import log from 'loglevel';
import { runSaga } from 'redux-saga';
import axios from 'axios';

const mockPersistentStore = (getData, setData) => {
  PersistentStore.get = jest.fn(() => getData);
  PersistentStore.set = jest.fn(() => setData);
  return PersistentStore;
};

const mockAxios = mockFunc => {
  axios.create = jest.fn().mockImplementation(() => ({
    post: mockFunc,
  }));
};
const dispatchedFunc = async (method, payload = {}) => {
  const dispatched = Array();
  let action = {};
  if (payload) action = { payload };
  await runSaga(
    {
      dispatch: action => dispatched.push(action),
    },
    method,
    action
  );
  return dispatched;
};

const mockisElectron = () => {
  return jest.spyOn(isElectron, 'isElectron');
};
const mockIpcRenderer = () => {
  return jest.spyOn(isElectron, 'ipcRendererFunc');
};

const spyLogger = () => {
  return jest.spyOn(log, 'error');
};

const mockNotification = () => {
  return (
    jest
      .spyOn(Notifications, 'default')
      // tslint:disable-next-line: no-empty
      .mockImplementation(() => {})
  );
};
export {
  mockPersistentStore,
  mockAxios,
  dispatchedFunc,
  mockisElectron,
  mockIpcRenderer,
  spyLogger,
  mockNotification,
};
