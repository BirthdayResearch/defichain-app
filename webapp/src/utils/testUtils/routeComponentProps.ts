import { createMemoryHistory, createLocation } from 'history';
import { match } from 'react-router';
const history = createMemoryHistory();
const path = `/route/:id`;
const match: match<{ id: string }> = {
  isExact: false,
  path,
  url: path.replace(':id', '1'),
  params: { id: '1' },
};

const location = createLocation(match.url);
delete history.location.key;
delete location.key;

export { history, match, location };
