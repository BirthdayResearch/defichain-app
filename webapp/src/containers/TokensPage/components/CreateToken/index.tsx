import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { RouteComponentProps } from 'react-router-dom';
import { TabContent, TabPane } from 'reactstrap';
import isEmpty from 'lodash/isEmpty';

import DCTDistribution from './DCTDistribution';
import CreateDCT from './CreateDCT';
import { createToken, fetchTokenInfo, updateToken } from '../../reducer';
import {
  CONFIRM_BUTTON_COUNTER,
  CONFIRM_BUTTON_TIMEOUT,
  CREATE_DCT,
  DCT_DISTRIBUTION,
  MINIMUM_DFI_REQUIRED_FOR_TOKEN_CREATION,
} from '../../../../constants';
import { ITokenResponse } from '../../../../utils/interfaces';
import { PaymentRequestModel } from '@defi_types/rpcConfig';
import { AddressModel } from '../../../../model/address.model';
import { getCountdownValue } from 'src/utils/utility';

interface RouteParams {
  id?: string;
}

interface CreateTokenProps extends RouteComponentProps<RouteParams> {
  tokenInfo: any;
  fetchToken: (id: string | undefined) => void;
  createToken: (tokenData) => void;
  updateToken: (tokenData) => void;
  createdTokenData: ITokenResponse;
  updatedTokenData: ITokenResponse;
  isTokenUpdating: boolean;
  isErrorUpdatingToken: string;
  isTokenCreating: boolean;
  isErrorCreatingToken: string;
  paymentRequests: PaymentRequestModel[];
}

export interface CreateTokenFormState extends AddressModel {
  [key: string]: any;
}

const CreateToken: React.FunctionComponent<CreateTokenProps> = (
  props: CreateTokenProps
) => {
  const { id } = props.match.params;
  const [activeTab, setActiveTab] = useState<string>(CREATE_DCT);
  const [
    IsCollateralAddressValid,
    setIsCollateralAddressValid,
  ] = useState<boolean>(true);
  const [formState, setFormState] = useState<CreateTokenFormState>({
    name: '',
    symbol: '',
    isDAT: false,
    decimal: '8',
    limit: '0',
    mintable: 'true',
    tradeable: 'true',
    receiveAddress: '',
    receiveLabel: '',
  });
  const [
    isConfirmationModalOpen,
    setIsConfirmationModalOpen,
  ] = useState<string>('default');
  const [wait, setWait] = useState<number>(5);
  const [allowCalls, setAllowCalls] = useState<boolean>(false);
  const [
    IsVerifyingCollateralModalOpen,
    setIsVerifyingCollateralModalOpen,
  ] = useState<boolean>(false);
  const [csvData, setCsvData] = React.useState<any>([]);

  const { fetchToken, tokenInfo } = props;

  useEffect(() => {
    fetchToken(id);
  }, []);

  const {
    createToken,
    updateToken,
    updatedTokenData,
    createdTokenData,
    isTokenCreating,
    isErrorCreatingToken,
    isTokenUpdating,
    isErrorUpdatingToken,
    paymentRequests,
  } = props;

  useEffect(() => {
    if (!isEmpty(tokenInfo) && id) {
      const data: CreateTokenFormState = {
        name: tokenInfo.name,
        symbol: tokenInfo.symbol,
        isDAT: tokenInfo.isDAT,
        decimal: tokenInfo.decimal.toString(),
        limit: tokenInfo.limit.toString(),
        mintable: tokenInfo.mintable.toString(),
        tradeable: tokenInfo.tradeable.toString(),
        receiveAddress: (paymentRequests ?? [])[0]?.address,
        receiveLabel: (paymentRequests ?? [])[0]?.label,
      };
      setFormState(data);
    }
  }, [tokenInfo]);

  useEffect(() => {
    async function addressAndAmount() {
      setFormState({
        ...formState,
        receiveAddress: (paymentRequests ?? [])[0]?.address,
        receiveLabel: (paymentRequests ?? [])[0]?.label,
      });
    }
    addressAndAmount();
  }, []);

  useEffect(() => {
    if (allowCalls && !isTokenCreating) {
      if (!isErrorCreatingToken && !isEmpty(createdTokenData)) {
        setIsConfirmationModalOpen('success');
      }
      if (isErrorCreatingToken && isEmpty(createdTokenData)) {
        setIsConfirmationModalOpen('failure');
      }
    }
  }, [createdTokenData, isTokenCreating, isErrorCreatingToken, allowCalls]);

  useEffect(() => {
    if (allowCalls && !isTokenUpdating) {
      if (!isErrorUpdatingToken && !isEmpty(updatedTokenData)) {
        setIsConfirmationModalOpen('success');
      }
      if (isErrorUpdatingToken && isEmpty(updatedTokenData)) {
        setIsConfirmationModalOpen('failure');
      }
    }
  }, [updatedTokenData, isTokenUpdating, isErrorUpdatingToken, allowCalls]);

  useEffect(() => {
    let waitToSendInterval;
    if (isConfirmationModalOpen === 'confirm') {
      let counter = CONFIRM_BUTTON_COUNTER;
      waitToSendInterval = setInterval(() => {
        counter -= 1;
        setWait(counter);
        if (counter === 0) {
          clearInterval(waitToSendInterval);
        }
      }, CONFIRM_BUTTON_TIMEOUT);
    }
    return () => {
      clearInterval(waitToSendInterval);
    };
  }, [isConfirmationModalOpen]);

  const handleChange = (e) => {
    setFormState({
      ...formState,
      [e.target.name]: e.target.value,
    });
  };

  const handleDropDowns = (newData: any, amount: any) => {
    if (amount < MINIMUM_DFI_REQUIRED_FOR_TOKEN_CREATION) {
      setIsCollateralAddressValid(false);
    } else {
      setFormState({
        ...formState,
        ...newData,
      });
      setIsCollateralAddressValid(true);
    }
  };

  const handleActiveTab = (active: string) => {
    setActiveTab(active);
  };

  const createTokenData = () => {
    return { ...formState, collateralAddress: formState.receiveAddress };
  };

  const cancelConfirmation = () => {
    setWait(5);
    setIsConfirmationModalOpen('default');
  };

  const createConfirmation = () => {
    setAllowCalls(true);
    const tokenData = createTokenData();
    setIsConfirmationModalOpen('loading');
    createToken(tokenData);
  };

  const updateConfirmation = () => {
    setAllowCalls(true);
    const tokenData = createTokenData();
    updateToken(tokenData);
  };

  const handleSubmit = async () => {
    const tokenData = createTokenData();
    createToken(tokenData);
  };

  useEffect(() => {
    if (!getCountdownValue()) {
      setWait(0);
    }
  });

  return (
    <TabContent activeTab={activeTab}>
      <TabPane tabId={CREATE_DCT}>
        <div className='main-wrapper position-relative'>
          <CreateDCT
            isUpdate={!isEmpty(tokenInfo) && !!id}
            handleChange={handleChange}
            formState={formState}
            isErrorCreatingToken={isErrorCreatingToken}
            createdTokenData={createdTokenData}
            updatedTokenData={updatedTokenData}
            isErrorUpdatingToken={isErrorUpdatingToken}
            wait={wait}
            setWait={setWait}
            createConfirmation={createConfirmation}
            updateConfirmation={updateConfirmation}
            cancelConfirmation={cancelConfirmation}
            IsCollateralAddressValid={IsCollateralAddressValid}
            isConfirmationModalOpen={isConfirmationModalOpen}
            setIsConfirmationModalOpen={setIsConfirmationModalOpen}
            handleDropDowns={handleDropDowns}
          />
        </div>
      </TabPane>
      <TabPane tabId={DCT_DISTRIBUTION}>
        <div className='main-wrapper position-relative'>
          <DCTDistribution
            setIsVerifyingCollateralModalOpen={
              setIsVerifyingCollateralModalOpen
            }
            IsVerifyingCollateralModalOpen={IsVerifyingCollateralModalOpen}
            handleActiveTab={handleActiveTab}
            csvData={csvData}
            setCsvData={setCsvData}
            handleSubmit={handleSubmit}
          />
        </div>
      </TabPane>
    </TabContent>
  );
};

const mapStateToProps = (state) => {
  const { tokens, wallet } = state;
  return {
    tokenInfo: tokens.tokenInfo,
    isTokenCreating: tokens.isTokenCreating,
    createdTokenData: tokens.createdTokenData,
    isErrorCreatingToken: tokens.isErrorCreatingToken,
    isTokenUpdating: tokens.isTokenUpdating,
    updatedTokenData: tokens.updatedTokenData,
    isErrorUpdatingToken: tokens.isErrorUpdatingToken,
    paymentRequests: wallet.paymentRequests,
  };
};

const mapDispatchToProps = {
  fetchToken: (id) => fetchTokenInfo({ id }),
  createToken: (tokenData) => createToken({ tokenData }),
  updateToken: (tokenData) => updateToken({ tokenData }),
};

export default connect(mapStateToProps, mapDispatchToProps)(CreateToken);
