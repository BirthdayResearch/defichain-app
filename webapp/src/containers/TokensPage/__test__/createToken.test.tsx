import React from 'react';
import CreateToken from '../components/CreateToken';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import store from '../../../app/rootStore';
import { match } from 'react-router';
import * as routeComponentProps from '../../../utils/testUtils/routeComponentProps';

describe('Create token page', () => {
  it('should check for snapshot', () => {
    const path = `/route/:hash`;
    const match: match<{ hash: string }> = {
      isExact: false,
      path,
      url: path.replace(':hash', '1'),
      params: { hash: '1' },
    };
    const ITokenResponse = {
      hash: 'hash',
    };
    const PaymentRequestModel = {
      label: 'label',
      id: '12',
      time: 'time',
      address: 'address',
      message: 'message',
      amount: '123',
      unit: '1',
      hdSeed: true,
    };
    const props = {
      tokenInfo: {},
      fetchToken: () => {},
      createToken: () => {},
      updateToken: () => {},
      createdTokenData: ITokenResponse,
      updatedTokenData: ITokenResponse,
      isTokenUpdating: true,
      isErrorUpdatingToken: 'isErrorUpdatingToken',
      isTokenCreating: true,
      isErrorCreatingToken: 'isErrorCreatingToken',
      paymentRequests: PaymentRequestModel,
      match: match,
    };
    const wrapper = shallow(
      <Provider store={store}>
        <CreateToken {...props} {...routeComponentProps} />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
