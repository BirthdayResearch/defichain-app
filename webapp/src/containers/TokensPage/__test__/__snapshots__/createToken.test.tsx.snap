// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create token page should check for snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Connect(CreateToken)
    createToken={[Function]}
    createdTokenData={
      Object {
        "hash": "hash",
      }
    }
    fetchToken={[Function]}
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
    isErrorCreatingToken="isErrorCreatingToken"
    isErrorUpdatingToken="isErrorUpdatingToken"
    isTokenCreating={true}
    isTokenUpdating={true}
    location={
      Object {
        "hash": "",
        "pathname": "/route/1",
        "search": "",
        "state": undefined,
      }
    }
    match={
      Object {
        "isExact": false,
        "params": Object {
          "id": "1",
        },
        "path": "/route/:id",
        "url": "/route/1",
      }
    }
    paymentRequests={
      Object {
        "address": "address",
        "amount": "123",
        "hdSeed": true,
        "id": "12",
        "label": "label",
        "message": "message",
        "time": "time",
        "unit": "1",
      }
    }
    tokenInfo={Object {}}
    updateToken={[Function]}
    updatedTokenData={
      Object {
        "hash": "hash",
      }
    }
  />
</ContextProvider>
`;
